server-common/src/main/java/org/apache/kafka/server/mutable/BoundedList.java

```diff
@@ -35,24 +35,12 @@ public class BoundedList<E> implements List<E> {
     private final int maxLength;
     private final List<E> underlying;
 
-    public static <E> BoundedList<E> newArrayBacked(int maxLength) {
-        return new BoundedList<>(maxLength, new ArrayList<>());
-    }
-
-    public static <E> BoundedList<E> newArrayBacked(int maxLength, int initialCapacity) {
```

Comment:
The operation which we can ensure the size of batch before processing can use this method.

Reply:
I'm not sure what you mean, could you please elaborate?

Reply:
> I'm not sure what you mean, could you please elaborate?

sorry for unclear comment. 

It seems to me "initialCapacity" is still useful if we "expect" the size of batch could be large. For example: create/delete a bunch of topics - that is a common usage. For such use cases, it would be better to set the initial capacity to avoid the resize problem you described.

Reply:
My understanding here is that we do want to know an expected size to start out with, but we shouldn't necessarily link it directly to an array we are passing in as the constructor does. Does that mean the deleted code here is actually the behavior we want?

Reply:
I think so. It looks like the consensus is:
1. Remove constructor that takes in a list.
2. Keep both versions of `newArrayBacked`, as ArrayList resizing operations don't seem to be a concern to people.

Reply:
1. maybe `private` is enough. the another view is that we can have `newListBacked` for more flexible usage in the future, and the implementation is based on the constructor taking a passed list.

```java
    public static <E> BoundedList<E> newListBacked(int maxLength) {
        return new BoundedList<>(maxLength, new LinkedList<>());
    }
```

2. "ArrayList resizing operations don't seem to be a concern to people." -> caller can fix it by defining "initialCapacity"

Reply:
I like the idea of migrating `newArrayBacked(int maxLength`) to `newLinkedListBacked(int maxLength)`, this would also help cut down on array list resizing time, as currently this constructor creates a default array list, which is size 10

Reply:
> I like the idea of migrating newArrayBacked(int maxLength) to newLinkedListBacked(int maxLength), this would also help cut down on array list resizing time, as currently this constructor creates a default array list, which is size 10

I did not suggest to use Linked list for current usage 🫢 

My point was keeping the constructor can have flexibility of changing the list implementation in the future 

We need to consider the cost of iteration before adopting the linked list.

To simplify this PR and address the origin propose, changing the scope from public to private is good enough. WDYT?

Reply:
Sure, sounds good to me

Reply:
Updated